meson.build

```diff
@@ -54,16 +54,27 @@ subdir('tests')
 install_data(licenses, install_dir: licensedir)
 
 pkg = import('pkgconfig')
-pkg.generate(
-  description: 'Library for managing JOSE objects',
-  version: meson.project_version(),
-  filebase: meson.project_name(),
-  name: 'José Library',
-
-  requires_private: [ 'zlib', 'libcrypto' ],
-  libraries: libjose_lib,
-  requires: 'jansson',
-)
+if host_machine.system() == 'freebsd'
+  pkg.generate(
+    description: 'Library for managing JOSE objects',
+    version: meson.project_version(),
+    filebase: meson.project_name(),
+    name: 'José Library',
+    requires_private: 'zlib',
```

Comment:
Shouldn't you just remove the quotes around zlib / libcrypto so that meson converts the dependency object itself into a pkg-config dependency? Why do you need to use different *string* values just for FreeBSD?

Reply:
If this is searching for the base system provided libraries, I’m not sure the pkg-config files get installed. I’ll check…

Reply:
And that's exactly why you want to use meson's rich object handling, because it can *tell* that a dependency wasn't *found* with pkg-config. So it won't add a pkg-config "requires".

(It will instead add a Libs linker flag, which is good, because you do still need to link to the library even on freebsd, even without a pkg-config file.)

Reply:
By the way, they started to provide them. See https://cgit.freebsd.org/src/commit/?id=b323fa85f15268ac44b8ff90faf90bce5d87b608

Newer versions of freebsd don't have this problem at all, and you want to make sure that on those systems, a proper pkg-config "requires" is added.

Reply:
Sure. As long as /usr/local/ has precedence over /usr, or can be configured in some way, I think relying on the png-config metadata is fine.

Reply:
I'm not certain what you mean by this since you're looking for the dependency either way already. My only advice here was to reuse the (successful) lookup that is saved in a variable, since meson's pkg-config module will automatically translate the dependency object to a guaranteed-correct pkg-config file, if you tell it that the generated pkg-config file should ***require the rich dependency object***.

Reply:
I'm not really wise to the way of meson; I'm more used to other build systems.
If there's a way to override the pkg-config somehow on a package by package basis, FreeBSD can manage picking whichever library is most appropriate via ports configuration options.

Reply:
This patch set has been pending for... a while.  It looks like the openssl changes went in last fall which post dates the patch and may make some of it unnecessary.  I'll try to find some time to poke at this and figure out what, if anything, is still necessary.

Reply:
I'll try and look at this some more, but a key is "in recent versions of FreeBSD".  It seems like the libcrypto pkg-config is only fixed in 13.2 and 14.  So 13.1 (EOL 7/31/23) and 12.4 (EOL 12/31/23) still need some workaround to find the default (i.e., not built in ports) openssl libraries.  Perhaps the best way forward is to leave this out of the patch, and continue to rely on the current, ugly patch in the ports tree which can then simply be removed in early 2024 without churn here.

Reply:
Sounds good to me at least. It might be a good idea to add some comments or file a proactive FreeBSD PR so it doesn’t get lost later..

Reply:
> If there's a way to override the pkg-config somehow on a package by package basis, FreeBSD can manage picking whichever library is most appropriate via ports configuration options.

Yes, the ports framework handles this. The underlying approach is that there's a variable respected by pkg-config, which works like $PATH does for binaries. It is called $PKG_CONFIG_PATH and permits adding most-preferred directories to search for .pc files. There's another variable if you want to completely ignore the system search path.

It's irrelevant to this review comment though, because...

> This patch set has been pending for... a while. It looks like the openssl changes went in last fall which post dates the patch and may make some of it unnecessary.

The openssl change is a "by the way" which I mentioned in a different comment from the comment where I explained why it's good to let meson handle this.

> I'll try and look at this some more, but a key is "in recent versions of FreeBSD". It seems like the libcrypto pkg-config is only fixed in 13.2 and 14. So 13.1 (EOL 7/31/23) and 12.4 (EOL 12/31/23) still need some workaround to find the default (i.e., not built in ports) openssl libraries. Perhaps the best way forward is to leave this out of the patch

Again, it really doesn't have anything to do with versions of freebsd. The current state of the PR doesn't include any code whatsoever to find the default libraries. And I'm not commenting about finding the libraries. All I'm commenting on is using meson's ***rich dependency objects*** as inputs to the pkg-config file generator.

Using meson's rich dependency objects as I puts her will always be guaranteed to do the right thing.
- if the dependency object is not-found, it does nothing
- if the dependency object was found because a recent freebsd version was used or the openssl port was used, and pkg-config was able to find it, then it generates a pkg-config "requires" field
- if the dependency object was found using some other detection method but not via pkg-config, perhaps via cc.find_library(), then it generates a pkg-config "Libs" field suitable for linking to dependencies that don't have pkg-config files

What this PR does instead:
- makes the pkg-config file erroneously say that it doesn't depend on any form of openssl-based software

tl;dr my advice is "good news, meson includes functionality to magically do this correctly no matter what. Just pass the return value of `dependency()` or `cc.find_library()` instead of quoted strings and it's guaranteed to work regardless of operating system or dependency type".

:)

Reply:
Ah, my apologies. I overlooked how the UX for this works in combination with the specific code used here... what's going on here is that:

```meson
requires_private: [libcrypto_pc],
libraries_private: [libcrypto_pc_or_library],
```

The requires_private field is a bit stricter than the libraries_private field. This is one of two differences between them, the other being "how a string value is handled".

You can test this with the following patch:
```diff
diff --git a/meson.build b/meson.build
index 96a2569..ca92669 100644
--- a/meson.build
+++ b/meson.build
@@ -34,10 +34,10 @@ add_project_arguments(
   language: 'c'
 )
 
-zlib = dependency('zlib')
+zlib = dependency('zlib', method: 'pkg-config')
 threads = dependency('threads')
 jansson = dependency('jansson', version: '>=2.10')
-libcrypto = dependency('libcrypto', version: '>=1.0.2')
+libcrypto = dependency('libcrypto', version: '>=1.0.2', method: 'system')
 a2x = find_program('a2x', required: get_option('docs'))
 jq = find_program('jq', required: false)
 
@@ -60,7 +60,7 @@ pkg.generate(
   filebase: meson.project_name(),
   name: 'José Library',
 
-  requires_private: [ 'zlib', 'libcrypto' ],
+  libraries_private: [ zlib, libcrypto ],
   libraries: libjose_lib,
   requires: 'jansson',
 )
```

Notice how I'm *forcing* libcrypto to use library fallback even on linux, and forcing zlib to use pkg-config (which is preferred anyway).

Here's the generated pkg-config file I get:
```
prefix=/usr
includedir=${prefix}/include
libdir=${prefix}/lib

Name: José Library
Description: Library for managing JOSE objects
Version: 11
Requires: jansson >= 2.10
Requires.private: zlib
Libs: -L${libdir} -ljose
Libs.private: -lcrypto -pthread
Cflags: -I${includedir}
```

Notice how:
- although I forced zlib to be found with pkg-config, it worked fine in libraries_private, and ended up in the Requires.private field
- although I forced libcrypto to be found with the "system" type (internally ignores pkg-config and runs its own find_library routines), the generator still handled it correctly and added it to Libs.private as `-lcrypto`.

Reply:
The thing is, you may not actually need to mention it at all. Here's what the pkg-config generator currently produces for the git master branch:
```
prefix=/usr
includedir=${prefix}/include
libdir=${prefix}/lib

Name: José Library
Description: Library for managing JOSE objects
Version: 11
Requires: jansson >= 2.10
Requires.private: zlib, libcrypto
Libs: -L${libdir} -ljose
Libs.private: -lcrypto -pthread
Cflags: -I${includedir}
```

Notice how it lists libcrypto twice, once in Requires.private and once in Libs.private.

If I ***comment out*** the requires_private line altogether, I get this:
```
prefix=/usr
includedir=${prefix}/include
libdir=${prefix}/lib

Name: José Library
Description: Library for managing JOSE objects
Version: 11
Requires: jansson >= 2.10
Requires.private: zlib
Libs: -L${libdir} -ljose
Libs.private: -lcrypto -pthread
Cflags: -I${includedir}
```

Meson has traced the dependencies of -ljose for you and determined it needs the `dependency('libcrypto')`.

Reply:
(It can be pretty useful to manually pass a dependency into the pkg-config generator, if you want to promote a dependency from private to public. Also probably when we rig up support for the pkgconf extension `Requires.internal`.)

Reply:
I've previously commented about this... https://github.com/mesonbuild/meson/issues/6037#issuecomment-543492958

Reply:
It looks like the libraries_private change is the right
answer.  I did not make changes to the depends method(s) since that seemed to make the Debian   
based builds very unhappy (couldn't find libcrypto) and everything seems to generate reasonable
pkg_config files and linker configs without invoking method.  THANK-YOU for your help.  I doubt I would have figured that out on my own.

Reply:
Right, my modifications of the method kwarg was just there to force Linux to act like FreeBSD and not find the pkg-config dependency.

As for the libraries_private change, I'm going to try to see if we can change the UX here, so that going forward, new users on newer meson versions don't get so confused here.